src/nodes/interceptor/eth_send_transaction.go

```diff
@@ -37,65 +47,183 @@ func (i *Interceptor) ethSendTransaction(ctx context.Context, msg *ethereum.Send
 		}
 		gas, err := sess.EthCaller().Eth().EstimateGas(ctx, callMsg)
 		if err != nil {
-			i.logger.WithError(err).Error("failed to estimate gas")
+			i.logger.WithError(err).With("gas_price", msg.GasPrice).Error("failed to estimate gas for quorum private transaction")
 			return nil, errors.BlockchainNodeError(err.Error())
 		}
 
 		msg.Gas = &gas
 	}
 
-	if msg.Nonce == nil {
-		n, err := sess.EthCaller().Eth().GetTransactionCount(ctx, msg.From, ethereum.PendingBlockNumber)
+	err := i.fillNonce(ctx, sess, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	if msg.Data == nil {
+		msg.Data = new([]byte)
+	}
+
+	var privateFrom string
+	if msg.PrivateFrom != nil {
+		privateFrom = *msg.PrivateFrom
+	}
+
+	// Store payload on Tessera
+	key, err := sess.ClientPrivTxManager().StoreRaw(ctx, *msg.Data, privateFrom)
+	if err != nil {
+		i.logger.WithError(err).Error("failed to store raw payload on Tessera", "private_from", privateFrom)
+		return nil, errors.BlockchainNodeError(err.Error())
+	}
+	// Switch message data
+	*msg.Data = key
+
+	raw, err := i.ethSignTransaction(ctx, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	hash, err := sess.EthCaller().Eth().SendRawPrivateTransaction(ctx, *raw, &msg.PrivateArgs)
+	if err != nil {
+		i.logger.WithError(err).Error("failed to send raw quorum private transaction")
+		return nil, errors.BlockchainNodeError(err.Error())
+	}
+
+	i.logger.Info("quorum private transaction sent successfully", "tx_hash", hash)
+	return &hash, nil
+}
+
+func (i *Interceptor) sendLegacyTx(ctx context.Context, msg *ethereum.SendTxMsg) (*ethcommon.Hash, error) {
+	i.logger.Debug("sending ETH legacy transaction")
+
+	sess := proxynode.SessionFromContext(ctx)
+
+	if msg.GasPrice == nil {
```

Comment:
I like the encapsulation you did with `fillNonce()`, could you do the same here?

Reply:
done